### PR TITLE
Send data

### PR DESCRIPTION
**Pull request by mayma18**
. add a message in the proto file to manage the sending of a string (sequence of bytes) to the master
. The slave extracts 1GB of data in a file generated by gensort (the file path is passed as an argument), and sends it to the master.
. The master checks that it knows the slaves, then extracts the data and adds it to its data field (List[Array[Byte]).

### CODE REVIEW

**Comment by mayma18** [comment]
add partition part

File: src/main/scala/cs332/distributedsorting/slave/Slave.scala [hunk 1]
```diff
@@ -1,51 +1,99 @@
...
+        // suppose args(1) is  the path to data file generated with gensort
+      val path = args.lastOption
+      if (path.isEmpty){
+        System.out.println("Path to data file is empty.")
+        client.shutdown()
+      }
+      else{
+        System.out.println("Try to send data to master")
+        val fSource = Source.fromFile(args(0))
+        val keyList = (fSource.grouped(100).map(x=>x.dropRight(90))).take(100000).toString() // list of key (size is 1GB)
```
Comment (dragon0170) [inline src/main/scala/cs332/distributedsorting/slave/Slave.scala]: I think 1GB is too big for each file. Maybe it will be enough for 1MB per file for sampling data.

**Review by dragon0170 — COMMENTED** [review]
I think the logic of sampling and partitioning key ranges is reasonable. good
Did you test the master and slave application with real gensort data?

File: src/main/scala/cs332/distributedsorting/slave/Slave.scala (hunk 1 above)
Comment (mayma18) [reply src/main/scala/cs332/distributedsorting/slave/Slave.scala]: I did not test with real gensort data yet. I will add this test to my weekly task. And for the size of sampling data, I change for 1MB, but if it 's not enough it will be very easy to increase it after the tests.

Comment (dragon0170) [reply src/main/scala/cs332/distributedsorting/slave/Slave.scala]: ok, than you can add commits for the test code of creating partition in this pull request.

File: src/main/protobuf/sorting.proto [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+package cs332.distributedsorting;
+
+service Sorting {
+  rpc Handshake (HandshakeRequest) returns (HandshakeResponse) {}
+  rpc SendData(SendDataRequest)returns (SendDataResponse){}
+}
+
+
+message SendDataRequest{
+  string data = 1;
```
Comment (dragon0170) [inline src/main/protobuf/sorting.proto]: protobuf supports `bytes` type. How about using `bytes` instead of `string`?
https://developers.google.com/protocol-buffers/docs/proto3#scalar

File: src/main/scala/cs332/distributedsorting/master/Master.scala [hunk 3]
```diff
@@ -45,12 +66,68 @@ class Master(executionContext: ExecutionContext) { self =>
...
+    val mindata = data.min(KeyOrdering)
+    val maxdata = data.max(KeyOrdering)
+    val range = div(count, sub(maxdata,mindata))
+    var loop  = 0
+    for (slave <- slaves ){
+      partition += (slave.toString ->(add(mindata, mult(loop,range)), add(mindata, mult(loop+1,range))))
+      loop+=1
+    }
+    return partition
+  }
```
Comment (dragon0170) [inline src/main/scala/cs332/distributedsorting/master/Master.scala]: I think that we have to consider the distribution of sampled key for deciding the key ranges.

Every key is 10 byte long, so it means a key can range over from 0 to (2^8)^10 = 2^80. So we should make our application to handle all the keys between 0 to 2^80 for sorting.

If we have `n` slaves, we should count the length of all sampled keys and divide them into `n` equal length sorted chunks. The last key of each chunk will be the upper bound for each `n` partition.
Also there are exceptional cases. The first partition must have lowerbound of 0 and the last partition must have upperbound of 2^80.

File: src/main/scala/cs332/distributedsorting/master/Master.scala [hunk 4]
```diff
@@ -67,29 +67,40 @@ class Master(executionContext: ExecutionContext, val numClient: Int) { self =>
...
   private def createPartition(): Map[String, (Array[Byte], Array[Byte])] = {
-    val mindata = data.min(KeyOrdering)
-    val maxdata = data.max(KeyOrdering)
-    val range = div(count, sub(maxdata,mindata))
-    var loop  = 0
-    for (slave <- slaves ){
-      partition += (slave.toString ->(add(mindata, mult(loop,range)), add(mindata, mult(loop+1,range))))
-      loop+=1
+    val mindata : Array[Byte] = Array(0,0,0,0,0,0,0,0,0,0)
+    val maxdata : Array[Byte] = Array(-2^7,-2^7,-2^7,-2^7,-2^7,-2^7,-2^7,-2^7,-2^7,-2^7)
```
Comment (dragon0170) [inline src/main/scala/cs332/distributedsorting/master/Master.scala]: I think minData is `Array(-128, -128, ..., -128)` and maxData is `Array(127, 127, ..., 127)`.
How did you choose the minData and maxData?

**Comment by mayma18** [comment]
fix issues we talk during the meeting. But it needs to implement test code.

File: src/main/scala/cs332/distributedsorting/master/Master.scala [hunk 5]
```diff
@@ -67,36 +69,39 @@ class Master(executionContext: ExecutionContext, val numClient: Int) { self =>
...
       }
-      if (loop == count - 1){
-        this.partition.put(slave.toString(), (data(loop*range), maxdata))
+      else if (loop == count-1){
+        this.partition.put(slave.toString(), (data((loop)*range -1), maxdata))
       }
       else{
-        this.partition.put(slave.toString(), (data(loop*range), data((loop+1)*range)))
+        this.partition.put(slave.toString(), (data(loop*range -1), data((loop+1)*range -1)))
       }
```
Comment (dragon0170) [inline src/main/scala/cs332/distributedsorting/master/Master.scala]: Why did you add -1 for every lowerbound and upperbound?
Is the lowerbound and upperbound is included or excluded? (ex. lowerbound: 3, upperbound: 7. It means 3 <= x <= 7 if all bounds mean included.)

Comment (mayma18) [reply src/main/scala/cs332/distributedsorting/master/Master.scala]: lower bound and upperbound are included that's why I add -1 to upperbound to not overlapp with the lowerbound of the next partition

File: src/main/scala/cs332/distributedsorting/slave/Slave.scala [hunk 6]
```diff
@@ -57,7 +58,8 @@ object Slave {
 
 class Slave private(
   private val channel: ManagedChannel,
-  private val blockingStub: SortingBlockingStub
+  private val blockingStub: SortingBlockingStub,
+  val name :String
```
Comment (mayma18) [inline src/main/scala/cs332/distributedsorting/slave/Slave.scala]: for debugging purpose I add a parameter name to Slave. Because basically Master use ip adress to distinguish two differents slaves. But in local, ip address are the same for all slaves. So instead I use a new parameter name

File: src/main/scala/cs332/distributedsorting/master/Master.scala (hunk 5 above)
Comment (dragon0170) [reply src/main/scala/cs332/distributedsorting/master/Master.scala]: But there is -1 in the lowerbound variable either. Is it intended?

Comment (mayma18) [reply src/main/scala/cs332/distributedsorting/master/Master.scala]: yes my mistake I changed it

File: src/main/scala/cs332/distributedsorting/master/Master.scala (hunk 4 above)
Comment (mayma18) [reply src/main/scala/cs332/distributedsorting/master/Master.scala]: It depends if we considere signed or unsigned bytes. I can't find any information about it in project presentation

Comment (dragon0170) [reply src/main/scala/cs332/distributedsorting/master/Master.scala]: Change to Byte.MinValue, Byte.MaxValue
